src/coreclr/System.Private.CoreLib/src/System/Threading/Interlocked.CoreCLR.cs

```diff
@@ -127,6 +128,26 @@ public static int CompareExchange(ref int location1, int value, int comparand)
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern int CompareExchange32(ref int location1, int value, int comparand);
 
+        // This is used internally in cases where having a managed
+        // ref to the location is unsafe (Ex: it is the object header of a pinned object).
+        // The intrinsic expansion for this overload is exactly the same as for the `ref int`
+        // variant and will go on the same path since expansion is triggered by the name and
+        // return type of the method.
+        // The important part is avoiding `ref *location` that is reported as byref to the GC.
+        [Intrinsic]
+        internal static unsafe int CompareExchange(int* location1, int value, int comparand)
+        {
+#if TARGET_X86 || TARGET_AMD64 || TARGET_ARM64 || TARGET_RISCV64
+            return CompareExchange(location1, value, comparand); // Must expand intrinsic
+#else
+            Debug.Assert(location1 != null);
+            return CompareExchange32Pointer(location1, value, comparand);
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.InternalCall)]
+        private static unsafe extern int CompareExchange32Pointer(int* location1, int value, int comparand);
```

Comment:
osx-arm64 build is failing, fix: https://github.com/dotnet/runtime/pull/122753.

Reply:
Analyzers should run for all build flavors by default. I do not think we have any conditions that enable them for debug builds only.

Also, I am not able to reproduce the error on Windows (tried `build -s clr.corelib -c debug` locally). I am not able to reproduce the error locally even when with `warnaserror+:IDE0036` passed directly on the csc command line when building CoreLib. I am not sure what's going on.